lib/disposable.dart

```diff
@@ -179,63 +182,86 @@ class Disposable implements _Disposable, DisposableManager {
     }
     _isDisposing = true;
 
-    List<Future> futures = []
-      ..addAll(_internalDisposables.map((disposable) => disposable.dispose()))
-      ..add(onDispose());
-
-    _internalDisposables = [];
-
+    await Future.wait(_blockingFutures);
     // We need to filter out nulls because a subscription cancel
     // method is allowed to return a plain old null value.
-    return Future
-        .wait(futures.where((future) => future != null))
-        .then(_completeDisposeFuture);
+    await Future.wait(_internalDisposables
+        .map((disposable) => disposable.dispose())
+        .where((future) => future != null));
+
+    _internalDisposables = null;
+    _blockingFutures = null;
+
+    return onDispose().then(_completeDisposeFuture);
   }
 
   /// Automatically dispose another object when this object is disposed.
   ///
   /// The parameter may not be `null`.
   @mustCallSuper
   @override
-  void manageDisposable(Disposable disposable) {
-    _throwIfNull(disposable, 'disposable');
+  Disposable manageDisposable(Disposable disposable) {
+    _throwOnInvalidCall(disposable, 'disposable');
     _internalDisposables.add(disposable);
+    return disposable;
   }
 
   /// Automatically handle arbitrary disposals using a callback.
   ///
   /// The parameter may not be `null`.
   @mustCallSuper
   @override
-  void manageDisposer(Disposer disposer) {
-    _throwIfNull(disposer, 'disposer');
+  void manageDisposer<T>(Disposer disposer) {
+    _throwOnInvalidCall(disposer, 'disposer');
     _internalDisposables.add(new _InternalDisposable(disposer));
   }
 
+  /// Causes the object to add the future to a list of futures which will
+  /// be awaited before the object disposes itself or any of the objects
+  /// under its management.
+  @mustCallSuper
+  @override
+  Future<T> manageFuture<T>(Future<T> future) {
+    _throwOnInvalidCall(future, 'manageFuture');
+    if (!_blockingFutures.contains(future)) {
+      Future removeFuture;
+      removeFuture = future.then((_) {
+        _blockingFutures.remove(future);
+        _blockingFutures.remove(removeFuture);
```

Comment:
~~Is it really necessary to remove this? Why not just use the future itself?~~

Okay, I think I see now. This tripped me up a bit, though; might be good to add a code comment.

I assume you're not using just the future itself so that you can "release" it as soon as possible, is that correct?